src/v/transform/transform_processor.cc

```diff
@@ -230,41 +244,62 @@ ss::future<> processor::run_transform_loop() {
         co_await _engine->transform(
           std::move(*batch),
           _probe,
-          [&transformed](
+          [this](
             std::optional<model::topic_view> topic,
             model::transformed_data data) {
-              vassert(topic == std::nullopt, "not supported yet 🙂");
-              transformed.push_back(std::move(data));
-              return ss::make_ready_future<wasm::write_success>(
-                wasm::write_success::yes);
+              auto output_topic = topic.value_or(
+                model::topic_view(_meta.output_topics.front().tp));
+              auto it = _outputs.find(output_topic);
+              if (it == _outputs.end()) {
+                  return ssx::now(wasm::write_success::no);
+              }
+              return it->second.queue.push({std::move(data)}, &_as).then([] {
+                  return wasm::write_success::yes;
+              });
           });
         vlog(_logger.trace, "transformed offset {}", offset);
-        if (!transformed.empty()) {
-            auto b = model::transformed_data::make_batch(
-              model::timestamp::now(), std::move(transformed));
-            co_await _transform_producer_pipe.push({std::move(b)}, &_as);
+        // Mark all queues as processsed up to this point.
+        for (auto& [_, output] : _outputs) {
+            co_await output.queue.push({offset}, &_as);
         }
-        co_await _transform_producer_pipe.push({offset}, &_as);
     }
 }
 
-ss::future<> processor::run_producer_loop() {
+ss::future<> processor::run_all_producers() {
+    std::vector<ss::future<>> futures;
+    for (auto& [topic, output] : _outputs) {
+        futures.push_back(run_producer_loop(&output.queue, output.sink.get()));
+    }
+    return ss::when_all_succeed(
+      std::make_move_iterator(futures.begin()),
+      std::make_move_iterator(futures.end()));
```

Comment:
fwiw, seastar already moves from begin/end

Reply:
Thanks, I pushed a commit to clean this up: https://github.com/redpanda-data/redpanda/pull/16440/commits/07dc8ffd171f5b8a2ca425dada77bc808e8bdf55